File: front-end/src/service/voucher-service.ts
import { AxiosResponse } from "axios";
import http from "@/service/http";
import { VoucherRequest, VoucherResponse } from "@/type/voucher-type";

// ================= API CHO NGƯỜI DÙNG =================

export const validateVoucher: (
    code: string,
    orderAmount: number
) => Promise<AxiosResponse<VoucherResponse>> = async (
    code,
    orderAmount
): Promise<AxiosResponse<VoucherResponse>> => {
    return http.get<VoucherResponse>("/vouchers/validate", {
        params: { code, orderAmount },
    });
};

// ================= API CHO ADMIN =================

export const getAllVouchers: () => Promise<
    AxiosResponse<VoucherResponse[]>
> = async (): Promise<AxiosResponse<VoucherResponse[]>> => {
    return http.get<VoucherResponse[]>("/vouchers");
};

export const getVoucherById: (
    id: number
) => Promise<AxiosResponse<VoucherResponse>> = async (
    id: number
): Promise<AxiosResponse<VoucherResponse>> => {
    return http.get<VoucherResponse>(`/vouchers/${id}`);
};

export const createVoucher: (
    data: VoucherRequest
) => Promise<AxiosResponse<VoucherResponse>> = async (
    data: VoucherRequest
): Promise<AxiosResponse<VoucherResponse>> => {
    return http.post<VoucherResponse>("/vouchers", data);
};

export const updateVoucher: (
    id: number,
    data: VoucherRequest
) => Promise<AxiosResponse<VoucherResponse>> = async (
    id: number,
    data: VoucherRequest
): Promise<AxiosResponse<VoucherResponse>> => {
    return http.put<VoucherResponse>(`/vouchers/${id}`, data);
};

export const disableVoucher: (
    id: number,
    action: boolean
) => Promise<AxiosResponse<void>> = async (
    id: number,
    action: boolean
): Promise<AxiosResponse<void>> => {
    return http.delete<void>(`/vouchers/${id}`, { params: { action } });
};

File: back-end/src/main/java/project/back_end/service/VoucherService.java
package project.back_end.service;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import project.back_end.request.VoucherRequest.VoucherRequest;
import project.back_end.response.UserVoucherResponse;
import project.back_end.response.VoucherResponse;

@Service
public interface VoucherService {
    // Lưu voucher vào kho của người dùng khi họ nhấn "Lưu" hoặc "Thu thập"
    void collectVoucher(Long userId, String voucherCode);

    /// Lấy danh sách voucher chưa sử dụng để hiển thị lên Popover ở FE
    Page<UserVoucherResponse> getUserVoucherWallet(Long userId, Pageable pageable);

    // Kiểm tra và trả về thông tin Voucher nếu hợp lệ với đơn hàng hiện tại
    VoucherResponse validateVoucher(Long userId, String code, Double orderAmount);

    // Đánh dấu voucher đã được sử dụng sau khi đặt hàng thành công
    void markVoucherAsUsed(Long userId, String code);

    // Tạo mới một voucher
    VoucherResponse createVoucher(VoucherRequest request);

    // Lấy danh sách tất cả voucher trong hệ thống
    Page<VoucherResponse> getAllVouchers(Pageable pageable);

    // Lấy chi tiết một voucher theo ID
    VoucherResponse getVoucherById(Long id);

    // Cập nhật thông tin voucher
    VoucherResponse updateVoucher(Long id, VoucherRequest request);

    // Xóa voucher (Soft Delete)
    void disableVoucher(Long id, Boolean action);
}